### ISSUE TITLE
Bug in TLSv1.3 client simulations

### ISSUE THREAD

**Issue by dcooper16** [drwetter/testssl.sh#1165]
There is a bug in `client_simulation_sockets()` when it comes to handling TLSv1.3 connections in which the server sends a HelloRetryRequest.

After sending the initial ClientHello, `client_simulation_sockets()` calls:
```
resend_if_hello_retry_request "$tls_hello_ascii" "$cipher_list_2send" "$4" "$process_full"
```
A minor issue is that neither "$4" nor "$process_full" are defined ("$process_full" should just be replaced by "ephemeralkey"). The more serious problem is that `resend_if_hello_retry_request()` will basically just call `prepare_tls_clienthello()` to create the second ClientHello, and this second ClientHello may not look anything like the initial ClientHello.

I would rate this bug as minor for two reasons: (1) not many servers support TLSv1.3 at the moment and (2) I think it would be very uncommon for a TLSv1.3 server to respond to an initial ClientHello with a HelloRetryRequest rather than a ServerHello. However, this will need to be fixed at some point, which may require writing a new version of `resend_if_hello_retry_request()` that is specifically designed to take a provided ClientHello and modify it as required by a HelloRetryRequest.

**Comment by drwetter** [comment]
Hi David,

is there a host out there where this bug hits?

Dirk

**Comment by dcooper16** [comment]
> is there a host out there where this bug hits?

I just came across this by chance when working on #1159. I noticed it as a result of looking at the code, not from testing.

In the client simulation data, "Chrome 70 Win 10" sends a supported_groups extension that offers X25519, P-256, and P384, but it only sends an X25519 key in the key_share extension. So, that would seem to be a good candidate for a client simulation in which the server would response with a HelloRetryRequest. However, every real TLSv1.3 server that I tried supported X25519, so this bug would not hit any of them.

I can easily create a server that supports TLSv1.3 but not X25519, so that the server responds with a HelloRetryRequest to the "Chrome 70 Win 10" ClientHello. When I do that, I can see that the ClientHello that testssl.sh sends in response to the HelloRetryRequest differs greatly from the initial ClientHello, which is a violation of RFC 8446. However, the server (OpenSSL 1.1.1a) seems to ignore this problem and so the end result is still the same as if a correctly generated ClientHello had been sent. So, even here the bug doesn't really "hit."

Still, some servers may compare the initial ClientHello with the one sent in response to the HelloRetryRequest and terminate the connection if they differ in ways not allowed by RFC 8446 (Section 4.1.2), so there is still the possibility that this bug could cause an incorrect result with some server.
